core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -31,256 +31,207 @@
 
 package io.grpc.internal;
 
-import static com.google.common.base.Preconditions.checkNotNull;
-
 import com.google.common.base.Preconditions;
 
 import io.grpc.Attributes;
 import io.grpc.Metadata;
 import io.grpc.Status;
 
-import java.io.InputStream;
-import java.util.logging.Level;
-import java.util.logging.Logger;
+import javax.annotation.Nullable;
 
 /**
- * Abstract base class for {@link ServerStream} implementations.
- *
- * @param <IdT> the type of the stream identifier
+ * Abstract base class for {@link ServerStream} implementations. Extending classes only need to
+ * implement {@link #transportState()} and {@link #abstractServerStreamSink()}.
  */
-public abstract class AbstractServerStream<IdT> extends AbstractStream<IdT>
-    implements ServerStream {
-  private static final Logger log = Logger.getLogger(AbstractServerStream.class.getName());
-
-  /** Whether listener.closed() has been called. */
-  private boolean listenerClosed;
-  private ServerStreamListener listener;
-
-  private boolean headersSent = false;
+public abstract class AbstractServerStream extends AbstractStream2
+    implements ServerStream, MessageFramer.Sink {
   /**
-   * Whether the stream was closed gracefully by the application (vs. a transport-level failure).
+   * Only called from application thread. A separate interface just to avoid name
```

Comment:
I've done a pass through the docs, which just so happened to improve the docs here. I think it addresses the first part of your confusion.

> This line "Only called from application thread." implies that it is the transport that is calling it (regardless of what thread it happens on).

No, actually. The transport _can't_ run on this application thread, because it doesn't know what that thread is and has no way to coordinate with it. I don't quite understand why the language would imply it was the transport calling, but I still think this may be addressed by the rewording I did.

As an aside, we're going to need to enhance our nomenclature here soon because we are likely to end up having three threads in play.